syft/frameworks/torch/linalg/lr.py

```diff
@@ -284,3 +285,247 @@ def _compute_pvalues(self):
             self.pvalue_intercept = 2 * t.cdf(-abs(tstat_intercept), self._dgf)
         else:
             self.pvalue_intercept = None
+
+
+class DASH:
+    """
+    Distributed Association Scan Hammer (DASH) algorithm based on Jonathan Bloom's algorithm.
+    It uses Secured Multi-Party Computation at combine phase.
+    While the training is performed in SMPC, the final regression coefficients
+    are public at the end.
+
+    Reference: Section 2 of https://arxiv.org/abs/1901.09531
+
+    Args:
+        crypto_provider: a BaseWorker providing crypto elements for ASTs such as
+            Beaver triples
+        hbc_worker: The "Honest but Curious" BaseWorker
+        precision_fractional: precision chosen for FixedPrecisionTensors
+
+    Attributes:
+        coef: torch.Tensor of shape (n_features, ). Estimated coefficients for
+            DASH algorithm.
+        pvalue: numpy.array of shape (n_features, ). Two-sided p-value for a
+            hypothesis test whose null hypothesis is that the each coeff is zero.
+
+    """
+
+    def __init__(
+        self, crypto_provider: BaseWorker, hbc_worker: BaseWorker, precision_fractional: int = 6
+    ):
+
+        self.crypto_provider = crypto_provider
+        self.hbc_worker = hbc_worker
+        self.precision_fractional = precision_fractional
+
+    def fit(
+        self, X_ptrs: List[torch.Tensor], C_ptrs: List[torch.Tensor], y_ptrs: List[torch.Tensor]
+    ):
+
+        # Checking if the pointers are as expected
+        self._check_ptrs(X_ptrs, C_ptrs, y_ptrs)
+
+        # Check if each y is a 2-dim or 1-dim tensor, unsqueeze it if it's 1-dim
+        for i, y in enumerate(y_ptrs):
+            if len(y.shape) < 2:
+                y_ptrs[i] = y.unsqueeze(1)
+
+        self.workers = self._get_workers(X_ptrs)
+
+        # Computing aggregated pairwise dot products remotelly
+        XX_ptrs, Xy_ptrs, yy_ptrs, CX_ptrs, Cy_ptrs = self._remote_dot_products(
+            X_ptrs, C_ptrs, y_ptrs
+        )
+
+        # Compute remote QR decompositions
+        R_ptrs = self._remote_qr(C_ptrs)
+
+        # Secred share tensors between hbc_worker, crypto_provider and a random worker
+        # and compute aggregates. It corresponds to the Combine stage of DASH's algorithm
+        idx = random.randint(0, len(self.workers) - 1)
+        XX_shared = sum(self._share_ptrs(XX_ptrs, idx)).sum(dim=0)
```

Comment:
You doing the sum twice here

Reply:
If you look at the [notebook with DASH implementation](https://github.com/jbloom22/DASH/blob/master/dash.ipynb) I added to Bloom's repo at his request, you will see that when we build the XXs matrices privately in the Compress step, we do a sum over the rows and at the Combine step we sum the XXs.
Here I am just doing the commutated operation, first I sum the XXs, then I sum the result over the row.
So the sum here is not redundant, it's needed.

Reply:
Finally, I decided to do it in the same order as in the notebook implementation, so now the sum over the axis 0 is being done at the call of the method `_remote_dot_products`. I have double-checked and the results are the same, but it's better to be in line with the notebook implem.